File: app/src/main/java/com/veganbeauty/app/features/account/notification/NotificationPushHelper.java
package com.veganbeauty.app.features.account.notification;

import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.media.RingtoneManager;
import android.os.Build;

import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationManagerCompat;

import com.veganbeauty.app.MainActivity;
import com.veganbeauty.app.R;
import com.veganbeauty.app.data.local.ProfileSession;

public class NotificationPushHelper {

    public static void sendPushNotification(
            Context context,
            String id,
            String title,
            String content,
            String notificationType,
            String voucherCode,
            String orderId,
            String scheduleId
    ) {
        boolean isNotiEnabled = ProfileSession.isNotiEnabled(context);
        boolean systemNotificationsEnabled = NotificationManagerCompat.from(context).areNotificationsEnabled();
        if (!isNotiEnabled || !systemNotificationsEnabled) {
            return;
        }

        int notificationId = Math.abs(id.hashCode());

        boolean soundEnabled = ProfileSession.isSoundEnabled(context);
        boolean vibrateEnabled = ProfileSession.isVibrateEnabled(context);

        String channelId = "rootie_push_channel_" + (soundEnabled ? "s1" : "s0") + "_" + (vibrateEnabled ? "v1" : "v0") + "_h";
        NotificationManager notificationManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            String name = "Rootie Notifications";
            String descriptionText = "Kênh gửi thông báo từ Rootie";
            int importance = soundEnabled ? NotificationManager.IMPORTANCE_HIGH : NotificationManager.IMPORTANCE_LOW;

            NotificationChannel channel = new NotificationChannel(channelId, name, importance);
            channel.setDescription(descriptionText);
            if (!soundEnabled) {
                channel.setSound(null, null);
            }
            channel.enableVibration(vibrateEnabled);
            if (!vibrateEnabled) {
                channel.setVibrationPattern(new long[]{0L});
            }
            notificationManager.createNotificationChannel(channel);
        }

        String actionName = "skin_chat".equalsIgnoreCase(notificationType) ? "open_skin_chat" : "open_notification_list";
        Intent bodyIntent = new Intent(context, MainActivity.class);
        bodyIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
        bodyIntent.putExtra("extra_notification_action", actionName);

        int pendingIntentFlags = Build.VERSION.SDK_INT >= Build.VERSION_CODES.M ?
                (PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE) :
                PendingIntent.FLAG_UPDATE_CURRENT;

        PendingIntent bodyPendingIntent = PendingIntent.getActivity(
                context,
                notificationId,
                bodyIntent,
                pendingIntentFlags
        );

        NotificationCompat.Builder builder = new NotificationCompat.Builder(context, channelId)
                .setSmallIcon(R.drawable.ic_notification)
                .setContentTitle(title)
                .setContentText(content)
                .setAutoCancel(true)
                .setContentIntent(bodyPendingIntent);

        if (!soundEnabled) {
            builder.setSound(null);
            builder.setPriority(NotificationCompat.PRIORITY_LOW);
        } else {
            builder.setSound(RingtoneManager.getDefaultUri(RingtoneManager.TYPE_NOTIFICATION));
            builder.setPriority(NotificationCompat.PRIORITY_HIGH);
        }

        if (!vibrateEnabled) {
            builder.setVibrate(new long[]{0L});
        } else {
            builder.setDefaults(NotificationCompat.DEFAULT_VIBRATE);
        }

        String actionText = null;
        if (notificationType != null) {
            switch (notificationType.toLowerCase()) {
                case "voucher": actionText = "COPY MÃ"; break;
                case "order": actionText = "CHI TIẾT"; break;
                case "skin care": actionText = "CHĂM DA NGAY"; break;
                case "checkin": actionText = "ĐIỂM DANH"; break;
                case "schedule date": actionText = "XEM LỊCH HẸN"; break;
            }
        }

        if (actionText != null) {
            Intent actionIntent = new Intent(context, MainActivity.class);
            actionIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
            actionIntent.putExtra("extra_notification_action", "open_detail");
            actionIntent.putExtra("extra_notification_type", notificationType);
            actionIntent.putExtra("extra_voucher_code", voucherCode);
            actionIntent.putExtra("extra_order_id", orderId);
            actionIntent.putExtra("extra_schedule_id", scheduleId);

            PendingIntent actionPendingIntent = PendingIntent.getActivity(
                    context,
                    notificationId + 1,
                    actionIntent,
                    pendingIntentFlags
            );
            builder.addAction(0, actionText, actionPendingIntent);
        }

        Intent skipIntent = new Intent(context, NotificationActionReceiver.class);
        skipIntent.setAction("com.veganbeauty.app.ACTION_SKIP");
        skipIntent.putExtra("extra_notification_id", notificationId);

        PendingIntent skipPendingIntent = PendingIntent.getBroadcast(
                context,
                notificationId + 2,
                skipIntent,
                pendingIntentFlags
        );
        builder.addAction(0, "Bỏ qua", skipPendingIntent);

        try {
            notificationManager.notify(notificationId, builder.build());
        } catch (SecurityException e) {
            e.printStackTrace();
        }
    }

    public static void sendCommunityPushNotification(
            Context context,
            String id,
            String title,
            String content,
            String actionName
    ) {
        boolean isNotiEnabled = ProfileSession.isNotiEnabled(context);
        boolean systemNotificationsEnabled = NotificationManagerCompat.from(context).areNotificationsEnabled();
        if (!isNotiEnabled || !systemNotificationsEnabled) {
            return;
        }

        int notificationId = Math.abs(id.hashCode());
        boolean soundEnabled = ProfileSession.isSoundEnabled(context);
        boolean vibrateEnabled = ProfileSession.isVibrateEnabled(context);

        String channelId = "rootie_community_channel_" + (soundEnabled ? "s1" : "s0") + "_" + (vibrateEnabled ? "v1" : "v0") + "_h";
        NotificationManager notificationManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            String name = "Rootie Community Notifications";
            String descriptionText = "Kênh gửi thông báo cộng đồng từ Rootie";
            int importance = soundEnabled ? NotificationManager.IMPORTANCE_HIGH : NotificationManager.IMPORTANCE_LOW;

            NotificationChannel channel = new NotificationChannel(channelId, name, importance);
            channel.setDescription(descriptionText);
            if (!soundEnabled) {
                channel.setSound(null, null);
            }
            channel.enableVibration(vibrateEnabled);
            if (!vibrateEnabled) {
                channel.setVibrationPattern(new long[]{0L});
            }
            notificationManager.createNotificationChannel(channel);
        }

        Intent bodyIntent = new Intent(context, MainActivity.class);
        bodyIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
        bodyIntent.putExtra("extra_notification_action", actionName != null ? actionName : "open_community_notification_list");

        int pendingIntentFlags = Build.VERSION.SDK_INT >= Build.VERSION_CODES.M ?
                (PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE) :
                PendingIntent.FLAG_UPDATE_CURRENT;

        PendingIntent bodyPendingIntent = PendingIntent.getActivity(
                context,
                notificationId,
                bodyIntent,
                pendingIntentFlags
        );

        NotificationCompat.Builder builder = new NotificationCompat.Builder(context, channelId)
                .setSmallIcon(R.drawable.ic_notification)
                .setContentTitle(title)
                .setContentText(content)
                .setAutoCancel(true)
                .setContentIntent(bodyPendingIntent);

        if (!soundEnabled) {
            builder.setSound(null);
            builder.setPriority(NotificationCompat.PRIORITY_LOW);
        } else {
            builder.setSound(RingtoneManager.getDefaultUri(RingtoneManager.TYPE_NOTIFICATION));
            builder.setPriority(NotificationCompat.PRIORITY_HIGH);
        }

        if (!vibrateEnabled) {
            builder.setVibrate(new long[]{0L});
        } else {
            builder.setDefaults(NotificationCompat.DEFAULT_VIBRATE);
        }

        Intent viewIntent = new Intent(context, MainActivity.class);
        viewIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
        viewIntent.putExtra("extra_notification_action", "open_community_notification_list");

        PendingIntent viewPendingIntent = PendingIntent.getActivity(
                context,
                notificationId + 1,
                viewIntent,
                pendingIntentFlags
        );
        builder.addAction(0, "Xem ngay", viewPendingIntent);

        Intent skipIntent = new Intent(context, NotificationActionReceiver.class);
        skipIntent.setAction("com.veganbeauty.app.ACTION_SKIP");
        skipIntent.putExtra("extra_notification_id", notificationId);

        PendingIntent skipPendingIntent = PendingIntent.getBroadcast(
                context,
                notificationId + 2,
                skipIntent,
                pendingIntentFlags
        );
        builder.addAction(0, "Bỏ qua", skipPendingIntent);

        try {
            notificationManager.notify(notificationId, builder.build());
        } catch (SecurityException e) {
            e.printStackTrace();
        }
    }
}

File: app/src/main/java/com/veganbeauty/app/features/account/notification/NotificationPushHelper.kt
package com.veganbeauty.app.features.account.notification

import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.os.Build
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import com.veganbeauty.app.MainActivity
import com.veganbeauty.app.R
import com.veganbeauty.app.data.local.ProfileSession

object NotificationPushHelper {

    fun sendPushNotification(
        context: Context,
        id: String,
        title: String,
        content: String,
        notificationType: String,
        voucherCode: String? = null,
        orderId: String? = null,
        scheduleId: String? = null
    ) {
        // Only send if user has enabled notifications in setting AND system allows it
        val isNotiEnabled = ProfileSession.isNotiEnabled(context)
        val systemNotificationsEnabled = NotificationManagerCompat.from(context).areNotificationsEnabled()
        if (!isNotiEnabled || !systemNotificationsEnabled) {
            return
        }

        val notificationId = Math.abs(id.hashCode())

        val soundEnabled = ProfileSession.isSoundEnabled(context)
        val vibrateEnabled = ProfileSession.isVibrateEnabled(context)

        val channelId = "rootie_push_channel_${if (soundEnabled) "s1" else "s0"}_${if (vibrateEnabled) "v1" else "v0"}_h"
        val notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val name = "Rootie Notifications"
            val descriptionText = "Kênh gửi thông báo từ Rootie"
            val importance = if (soundEnabled) {
                NotificationManager.IMPORTANCE_HIGH
            } else {
                NotificationManager.IMPORTANCE_LOW
            }

            val channel = NotificationChannel(channelId, name, importance).apply {
                description = descriptionText
                if (!soundEnabled) {
                    setSound(null, null)
                }
                enableVibration(vibrateEnabled)
                if (!vibrateEnabled) {
                    vibrationPattern = longArrayOf(0L)
                }
            }
            notificationManager.createNotificationChannel(channel)
        }

        // 1. PendingIntent for body click: Opens MainActivity and routes to correct screen
        val actionName = if (notificationType.equals("skin_chat", ignoreCase = true)) "open_skin_chat" else "open_notification_list"
        val bodyIntent = Intent(context, MainActivity::class.java).apply {
            flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
            putExtra("extra_notification_action", actionName)
        }
        val pendingIntentFlags = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        } else {
            PendingIntent.FLAG_UPDATE_CURRENT
        }
        val bodyPendingIntent = PendingIntent.getActivity(
            context,
            notificationId,
            bodyIntent,
            pendingIntentFlags
        )

        // 2. Build Notification
        val builder = NotificationCompat.Builder(context, channelId)
            .setSmallIcon(R.drawable.ic_notification)
            .setContentTitle(title)
            .setContentText(content)
            .setAutoCancel(true)
            .setContentIntent(bodyPendingIntent)

        // Set Sound/Vibration
        if (!soundEnabled) {
            builder.setSound(null)
            builder.priority = NotificationCompat.PRIORITY_LOW
        } else {
            builder.setSound(android.media.RingtoneManager.getDefaultUri(android.media.RingtoneManager.TYPE_NOTIFICATION))
            builder.priority = NotificationCompat.PRIORITY_HIGH
        }
        if (!vibrateEnabled) {
            builder.setVibrate(longArrayOf(0L))
        } else {
            builder.setDefaults(NotificationCompat.DEFAULT_VIBRATE)
        }

        // 3. Action Button (e.g. COPY MÃ, CHI TIẾT, etc.)
        val actionText = when (notificationType.lowercase()) {
            "voucher" -> "COPY MÃ"
            "order" -> "CHI TIẾT"
            "skin care" -> "CHĂM DA NGAY"
            "checkin" -> "ĐIỂM DANH"
            "schedule date" -> "XEM LỊCH HẸN"
            else -> null
        }

        if (actionText != null) {
            val actionIntent = Intent(context, MainActivity::class.java).apply {
                flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
                putExtra("extra_notification_action", "open_detail")
                putExtra("extra_notification_type", notificationType)
                putExtra("extra_voucher_code", voucherCode)
                putExtra("extra_order_id", orderId)
                putExtra("extra_schedule_id", scheduleId)
            }
            val actionPendingIntent = PendingIntent.getActivity(
                context,
                notificationId + 1,
                actionIntent,
                pendingIntentFlags
            )
            builder.addAction(0, actionText, actionPendingIntent)
        }

        // 4. "Bỏ qua" Action Button
        val skipIntent = Intent(context, NotificationActionReceiver::class.java).apply {
            action = "com.veganbeauty.app.ACTION_SKIP"
            putExtra("extra_notification_id", notificationId)
        }
        val skipPendingIntent = PendingIntent.getBroadcast(
            context,
            notificationId + 2,
            skipIntent,
            pendingIntentFlags
        )
        builder.addAction(0, "Bỏ qua", skipPendingIntent)

        // Show Notification
        try {
            notificationManager.notify(notificationId, builder.build())
        } catch (e: SecurityException) {
            e.printStackTrace()
        }
    }

    @JvmStatic
    fun sendCommunityPushNotification(
        context: Context,
        id: String,
        title: String,
        content: String,
        actionName: String = "open_community_notification_list"
    ) {
        val isNotiEnabled = ProfileSession.isNotiEnabled(context)
        val systemNotificationsEnabled = NotificationManagerCompat.from(context).areNotificationsEnabled()
        if (!isNotiEnabled || !systemNotificationsEnabled) {
            return
        }

        val notificationId = Math.abs(id.hashCode())
        val soundEnabled = ProfileSession.isSoundEnabled(context)
        val vibrateEnabled = ProfileSession.isVibrateEnabled(context)

        val channelId = "rootie_community_channel_${if (soundEnabled) "s1" else "s0"}_${if (vibrateEnabled) "v1" else "v0"}_h"
        val notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val name = "Rootie Community Notifications"
            val descriptionText = "Kênh gửi thông báo cộng đồng từ Rootie"
            val importance = if (soundEnabled) {
                NotificationManager.IMPORTANCE_HIGH
            } else {
                NotificationManager.IMPORTANCE_LOW
            }

            val channel = NotificationChannel(channelId, name, importance).apply {
                description = descriptionText
                if (!soundEnabled) {
                    setSound(null, null)
                }
                enableVibration(vibrateEnabled)
                if (!vibrateEnabled) {
                    vibrationPattern = longArrayOf(0L)
                }
            }
            notificationManager.createNotificationChannel(channel)
        }

        val bodyIntent = Intent(context, MainActivity::class.java).apply {
            flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
            putExtra("extra_notification_action", actionName)
        }
        val pendingIntentFlags = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        } else {
            PendingIntent.FLAG_UPDATE_CURRENT
        }
        val bodyPendingIntent = PendingIntent.getActivity(
            context,
            notificationId,
            bodyIntent,
            pendingIntentFlags
        )

        val builder = NotificationCompat.Builder(context, channelId)
            .setSmallIcon(R.drawable.ic_notification)
            .setContentTitle(title)
            .setContentText(content)
            .setAutoCancel(true)
            .setContentIntent(bodyPendingIntent)

        if (!soundEnabled) {
            builder.setSound(null)
            builder.priority = NotificationCompat.PRIORITY_LOW
        } else {
            builder.setSound(android.media.RingtoneManager.getDefaultUri(android.media.RingtoneManager.TYPE_NOTIFICATION))
            builder.priority = NotificationCompat.PRIORITY_HIGH
        }
        if (!vibrateEnabled) {
            builder.setVibrate(longArrayOf(0L))
        } else {
            builder.setDefaults(NotificationCompat.DEFAULT_VIBRATE)
        }

        // Action Button: "Xem ngay"
        val viewIntent = Intent(context, MainActivity::class.java).apply {
            flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
            putExtra("extra_notification_action", "open_community_notification_list")
        }
        val viewPendingIntent = PendingIntent.getActivity(
            context,
            notificationId + 1,
            viewIntent,
            pendingIntentFlags
        )
        builder.addAction(0, "Xem ngay", viewPendingIntent)

        // "Bỏ qua" Action Button
        val skipIntent = Intent(context, NotificationActionReceiver::class.java).apply {
            action = "com.veganbeauty.app.ACTION_SKIP"
            putExtra("extra_notification_id", notificationId)
        }
        val skipPendingIntent = PendingIntent.getBroadcast(
            context,
            notificationId + 2,
            skipIntent,
            pendingIntentFlags
        )
        builder.addAction(0, "Bỏ qua", skipPendingIntent)

        try {
            notificationManager.notify(notificationId, builder.build())
        } catch (e: SecurityException) {
            e.printStackTrace()
        }
    }
}
